### PR TITLE
Limit ring hash max size to 4K

### PR DESCRIPTION
**Pull request by apolcyn**
Implements https://github.com/grpc/proposal/pull/338 for Java.

Note: a configuration option (preferable taking the form of a dial option, but possibly using a global) is left as a TODO.

cc @ejona86

### CODE REVIEW

**Comment by ejona86** [comment]
There's quite a few parts from grpc/proposal#338 missing, as the xds resource processing also has to be adjusted to pass down 8 MB when unset. That's mostly internal, so I think the biggest thing that is needed before merging is some sort of option to change the limit. That was the biggest concern with this change.

**Comment by apolcyn** [comment]
Sure we can add the option to this PR, but @ejona86 can you propose an API?

File: xds/src/main/java/io/grpc/xds/RingHashOptions.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import io.grpc.ExperimentalApi;
+
+/**
+ * Utility class that provides a way to configure ring hash size limits. This is applicable
+ * for clients that use the ring hash load balancing policy. Note that size limits involve
+ * a tradeoff between client memory consumption and accuracy of load balancing weight
+ * representations. Also see https://github.com/grpc/proposal/pull/338.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/9718")
+public final class RingHashOptions {
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/RingHashOptions.java:29]: Add a private default constructor, something like:
```java
  private RingHashOptions() {} // Prevent instantiation
```

File: xds/src/main/java/io/grpc/xds/RingHashOptions.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  static final long MAX_RING_SIZE_CAP = 8 * 1024 * 1024L;
+  @VisibleForTesting
+  // Same as RingHashLoadBalancerProvider.DEFAULT_MAX_RING_SIZE
+  static final long DEFAULT_RING_SIZE_CAP = 4 * 1024L;
+
+  // Limits ring hash sizes to restrict client memory usage.
+  private static volatile long ringSizeCap = DEFAULT_RING_SIZE_CAP;
+
+  /**
+   * Set the global limit for min and max ring hash sizes. Note that
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/RingHashOptions.java:41]: "the min and max number of ring hash entries."

File: xds/src/main/java/io/grpc/xds/RingHashOptions.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   * this limit is clamped between 1 and 8M, and attempts to set
+   * the limit lower or higher than that range will be silently
+   * moved to the nearest number within that range. Defaults initially
+   * to 4K.
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/RingHashOptions.java:45]: 4096

**Review by ejona86 — APPROVED** [review]
Some editorial stuff and nits, but otherwise LGTM

File: xds/src/main/java/io/grpc/xds/RingHashOptions.java (hunk 2 above)
Comment (apolcyn) [reply xds/src/main/java/io/grpc/xds/RingHashOptions.java]: done

File: xds/src/main/java/io/grpc/xds/RingHashOptions.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  @VisibleForTesting
+  // Same as RingHashLoadBalancerProvider.DEFAULT_MAX_RING_SIZE
+  static final long DEFAULT_RING_SIZE_CAP = 4 * 1024L;
+
+  // Limits ring hash sizes to restrict client memory usage.
+  private static volatile long ringSizeCap = DEFAULT_RING_SIZE_CAP;
+
+  /**
+   * Set the global limit for min and max ring hash sizes. Note that
+   * this limit is clamped between 1 and 8M, and attempts to set
```
[the remark this answers is not in the document]
Comment (apolcyn) [reply xds/src/main/java/io/grpc/xds/RingHashOptions.java]: updated wording to clarify

File: xds/src/main/java/io/grpc/xds/RingHashOptions.java (hunk 3 above)
Comment (apolcyn) [reply xds/src/main/java/io/grpc/xds/RingHashOptions.java]: done

File: xds/src/main/java/io/grpc/xds/RingHashOptions.java (hunk 1 above)
Comment (apolcyn) [reply xds/src/main/java/io/grpc/xds/RingHashOptions.java:29]: done

**Pull request merged**